core/src/main/java/bisq/core/dao/node/BsqNode.java

```diff
@@ -179,19 +189,87 @@ protected void onParseBlockChainComplete() {
         log.info("onParseBlockChainComplete");
         parseBlockchainComplete = true;
         daoStateService.onParseBlockChainComplete();
-
-        // log.error("COMPLETED: sb1={}\nsb2={}", BlockParser.sb1.toString(), BlockParser.sb2.toString());
-        // log.error("equals? " + BlockParser.sb1.toString().equals(BlockParser.sb2.toString()));
-        // Utilities.copyToClipboard(BlockParser.sb1.toString() + "\n\n\n" + BlockParser.sb2.toString());
     }
 
     @SuppressWarnings("WeakerAccess")
     protected void startReOrgFromLastSnapshot() {
         daoStateSnapshotService.applySnapshot(true);
-        startParseBlocks();
     }
 
-    protected boolean isBlockAlreadyAdded(RawBlock rawBlock) {
-        return daoStateService.getBlockAtHeight(rawBlock.getHeight()).isPresent();
+
+    protected Optional<Block> doParseBlock(RawBlock rawBlock) throws RequiredReorgFromSnapshotException {
+        // We check if we have a block with that height. If so we return. We do not use the chainHeight as with genesis
+        // height we have no block but chainHeight is initially set to genesis height (bad design ;-( but a bit tricky
+        // to change now as it used in many areas.)
+        if (daoStateService.getBlockAtHeight(rawBlock.getHeight()).isPresent()) {
+            log.info("We have already a block with the height of the new block. Height of new block={}", rawBlock.getHeight());
+            return Optional.empty();
+        }
+
+        try {
+            Block block = blockParser.parseBlock(rawBlock);
+
+            if (pendingBlocks.contains(rawBlock))
+                pendingBlocks.remove(rawBlock);
+
+            // After parsing we check if we have pending blocks we might have received earlier but which have been
+            // not connecting from the latest height we had. The list is sorted by height
+            if (!pendingBlocks.isEmpty()) {
+                // To avoid ConcurrentModificationException we copy the list. It might be altered in the method call
+                ArrayList<RawBlock> tempPendingBlocks = new ArrayList<>(pendingBlocks);
+                for (RawBlock tempPendingBlock : tempPendingBlocks) {
+                    try {
+                        doParseBlock(tempPendingBlock);
+                    } catch (RequiredReorgFromSnapshotException e1) {
```

Comment:
Catching the RequiredReorgFromSnapshotException here and braking will stop this loop but if there was a loop in a previous level of recursion that would continue running and parsing in a now known bad state. I think the exception has to be rethrown here.

Reply:
Tested a bit and did a 6 block reorg by generating 10 blocks, invalidating the fourth of those blocks and generating 10 more blocks. Results in almost 8200 useless iterations.

It also seems to have broken the phase/cycle handling. Created an issue about it, shouldn't have since it's probably due to the reorg handling, #2064

Reply:
Good point about the RequiredReorgFromSnapshotException. I think if we use the first item instead of loop it should be ok. I will look into the other issue. Have not tested how it effect other aspects. We should make a client with all features used (requests, voting, bonds,...) and then apply  a reorg to see if it break anything. Might be valid result if stuff become invalid by deep reorgs - e.g. < break, with < break it should not change results.